{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# User Inputs\n",
    "\n",
    "simulation_folder_names =  '../simulations/new_folder_' # Name scheme of the directories used to save simulations\n",
    "\n",
    "configs = 1 # number of replicate cells\n",
    "\n",
    "simtime = 1200 # total time simulated, equal to simulation_time in the program. Must be integer.\n",
    "\n",
    "## WARNING: Some pieces of code can takes several minutes to evaluate, the data files are large and processing \n",
    "##          genome-wide data takes some time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### General Naming Scheme ###\n",
    "\n",
    "# Proteins: P_gene number Examples: for gene JCVISYN3A_0002, the protein in P_2 for gene JCVISYN3A_0131, the protein is 0131\n",
    "        \n",
    "# mRNA: R_gene number\n",
    "    \n",
    "# RNAP bound to gene: RP_gene number\n",
    "    \n",
    "# mRNA bound to degradosome: D_gene number\n",
    "    \n",
    "# mRNA bound to ribosomes:\n",
    "    \n",
    "#     to single ribosome: RB_gene number\n",
    "        \n",
    "#     to start of a polysome: PS_gene number\n",
    "        \n",
    "#     to middle of a polysome: PM_gene number\n",
    "        \n",
    "#     to end of a polysome: PE_gene number\n",
    "        \n",
    "#     If mRNA is too short for a polysome:\n",
    "        \n",
    "#         RBS_, RBM_ and RBE_ instead of PS_, PM_, and PE_\n",
    "        \n",
    "# For special particle names, see files in model data:\n",
    "    \n",
    "#     protein_metabolites_frac.csv\n",
    "    \n",
    "#     ribo_protein_metabolites.csv\n",
    "    \n",
    "#     trna_metabolites_synthase.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from jLM.RegionBuilder import RegionBuilder\n",
    "from jLM.RDME import Sim as RDMESim\n",
    "from jLM.RDME import File as RDMEFile\n",
    "import jLM\n",
    "\n",
    "from pyLM.units import *\n",
    "\n",
    "import lm\n",
    "\n",
    "from lm import MpdRdmeSolver\n",
    "from lm import IntMpdRdmeSolver\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "import numpy  as np\n",
    "\n",
    "import os\n",
    "\n",
    "import scipy.ndimage as spnd\n",
    "import ipywidgets as ipw\n",
    "import h5py\n",
    "import itertools\n",
    "import random\n",
    "import copy\n",
    "\n",
    "# import ipyvolume\n",
    "# from sidecar import Sidecar\n",
    "# import numpy as np\n",
    "# from ipywebrtc import WidgetStream, VideoRecorder\n",
    "\n",
    "import math\n",
    "import scipy as sp\n",
    "import scipy.spatial\n",
    "\n",
    "# import seaborn as sns\n",
    "\n",
    "import csv\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import importlib\n",
    "from collections import defaultdict, OrderedDict\n",
    "\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "plt.rcParams.update({'font.size': 18})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# traj = RDMEFile('MinCell_jLM_RDME_CME_ODE_2DNA.lm',replicate=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "workDir = simulation_folder_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "Max_Poly_Count = 10\n",
    "ht = np.arange(0,simtime+1,1)\n",
    "ft = np.arange(1,int(simtime/60),1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Loading multiple trajectories can take several minutes\n",
    "\n",
    "trajs = []\n",
    "\n",
    "for rep in range(configs):\n",
    "\n",
    "    filename = workDir + str(rep+1) + '/MinCell_jLM_RDME_CME_ODE_polysomes.lm'\n",
    "    traj = RDMEFile(filename,replicate=1)\n",
    "    trajs.append(traj)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def avgHybridCnt(specID):\n",
    "    \n",
    "    avgParts = np.zeros(simtime+1)\n",
    "    \n",
    "    for i in range(configs):\n",
    "        PCDF = pd.read_csv(workDir + str(i+1) + '/particle_counts.csv')\n",
    "        partTrace = PCDF.loc[ PCDF.Time == specID ].values[0]\n",
    "    #     print(partTrace)\n",
    "        partTrace = np.delete(partTrace,0)\n",
    "    #     print(partTrace)\n",
    "        partTrace.astype('float64')\n",
    "        \n",
    "        avgParts = avgParts + partTrace\n",
    "    \n",
    "    avgParts = avgParts/configs\n",
    "    return avgParts\n",
    "\n",
    "def repHybridCnt(specID,rep):\n",
    "    \n",
    "    avgParts = np.zeros(simtime+1)\n",
    "    \n",
    "    for i in range(configs):\n",
    "        if i==rep:\n",
    "            PCDF = pd.read_csv(workDir + str(i+1) + '/particle_counts.csv')\n",
    "            partTrace = PCDF.loc[ PCDF.Time == specID ].values[0]\n",
    "        #     print(partTrace)\n",
    "            partTrace = np.delete(partTrace,0)\n",
    "        #     print(partTrace)\n",
    "            partTrace.astype('float64')\n",
    "\n",
    "            avgParts = avgParts + partTrace\n",
    "    \n",
    "#     avgParts = avgParts/configs\n",
    "    return avgParts\n",
    "\n",
    "def avgHybridConc(specID):\n",
    "    \n",
    "    avgConc = np.zeros(simtime+1)\n",
    "    \n",
    "    for i in range(configs):\n",
    "        concTrace = []\n",
    "        PCDF = pd.read_csv(workDir + str(i+1) + '/particle_counts.csv')\n",
    "#         print(workDir + str(i+1) + '/particle_counts.csv')\n",
    "        partTrace = PCDF.loc[ PCDF.Time == specID ].values[0]\n",
    "    #     print(partTrace)\n",
    "        partTrace = np.delete(partTrace,0)\n",
    "    #     print(partTrace)\n",
    "        partTrace.astype('float64')\n",
    "\n",
    "        SATrace = PCDF.loc[ PCDF.Time == 'CellSA' ].values[0]\n",
    "    #     print(partTrace)\n",
    "        SATrace = np.delete(SATrace,0)\n",
    "    #     print(partTrace)\n",
    "        SATrace.astype('float64')\n",
    "    \n",
    "        for i in range(len(partTrace)):\n",
    "\n",
    "            cellRadius = ((SATrace[i]/4/np.pi)**(1/2))*1e-9\n",
    "            cellVolume = ((4/3)*np.pi*(cellRadius)**3)*(1000)\n",
    "\n",
    "            conc = (partTrace[i]*1000.0)/((6.02e23)*cellVolume)\n",
    "\n",
    "            concTrace.append(conc)\n",
    "        \n",
    "        avgConc += concTrace\n",
    "#         print(avgConc)\n",
    "    avgConc = avgConc/configs\n",
    "    return avgConc\n",
    "\n",
    "def repHybridConc(specID,rep):\n",
    "    \n",
    "    avgConc = np.zeros(simtime+1)\n",
    "    \n",
    "    for i in range(configs):\n",
    "        if i==rep:\n",
    "            concTrace = []\n",
    "            PCDF = pd.read_csv(workDir + str(i+1) + '/particle_counts.csv')\n",
    "            partTrace = PCDF.loc[ PCDF.Time == specID ].values[0]\n",
    "        #     print(partTrace)\n",
    "            partTrace = np.delete(partTrace,0)\n",
    "        #     print(partTrace)\n",
    "            partTrace.astype('float64')\n",
    "\n",
    "            SATrace = PCDF.loc[ PCDF.Time == 'CellSA' ].values[0]\n",
    "        #     print(partTrace)\n",
    "            SATrace = np.delete(SATrace,0)\n",
    "        #     print(partTrace)\n",
    "            SATrace.astype('float64')\n",
    "\n",
    "            for i in range(len(partTrace)):\n",
    "\n",
    "                cellRadius = ((SATrace[i]/4/np.pi)**(1/2))*1e-9\n",
    "                cellVolume = ((4/3)*np.pi*(cellRadius)**3)*(1000)\n",
    "\n",
    "                conc = (partTrace[i]*1000.0)/((6.02e23)*cellVolume)\n",
    "\n",
    "                concTrace.append(conc)\n",
    "\n",
    "            avgConc = avgConc + concTrace\n",
    "#     avgConc = avgConc/configs\n",
    "    return avgConc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pltAvgRna(RnaName,RiboName,DegName):\n",
    "    \n",
    "    RNAcount = np.zeros(simtime+1)\n",
    "    \n",
    "    for i in range(configs):\n",
    "        \n",
    "        t, Rcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RnaName))\n",
    "        \n",
    "        RNAcount = RNAcount + Rcount\n",
    "    \n",
    "        t, RRcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboName))\n",
    "        \n",
    "        RNAcount = RNAcount + RRcount\n",
    "    \n",
    "        t, RDEGcount = trajs[i].getNumberTrajectory(species = trajs[i].species(DegName))\n",
    "    \n",
    "        RNAcount = RNAcount + RDEGcount\n",
    "    \n",
    "    RNAcount = RNAcount/configs\n",
    "    \n",
    "    return t, RNAcount\n",
    "\n",
    "def pltAvgSpec(specName):\n",
    "    \n",
    "    Count = np.zeros(simtime+1)\n",
    "    \n",
    "    for i in range(configs):\n",
    "        \n",
    "        t, Rcount = trajs[i].getNumberTrajectory(species = trajs[i].species(specName))\n",
    "        \n",
    "        Count = Count + Rcount\n",
    "    \n",
    "    Count = Count/configs\n",
    "    \n",
    "    return t, Count    \n",
    "\n",
    "\n",
    "def pltAvgPtn(specName):\n",
    "    \n",
    "    Count = np.zeros(simtime+1)\n",
    "    \n",
    "    for i in range(configs):\n",
    "        \n",
    "        t, Rcount = trajs[i].getNumberTrajectory(species = trajs[i].species(specName))\n",
    "        \n",
    "        Count = Count + Rcount\n",
    "    \n",
    "    Count = Count/configs\n",
    "    \n",
    "    return t, Count \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pltAvgRna(locusTag):\n",
    "    \n",
    "    locusNum = locusTag.split('_')[1].lstrip('0')\n",
    "    \n",
    "    rnasequence, aasequence = getSequences(locusTag)\n",
    "    \n",
    "    aaCount = defaultdict(int)\n",
    "    for aa in set(aasequence):\n",
    "        aaCount[aa] = aasequence.count(aa)\n",
    "    ptn_len = sum(list(aaCount.values()))\n",
    "\n",
    "    if ptn_len <= 125:\n",
    "        max_poly_size = 1\n",
    "    elif ptn_len > 125:\n",
    "        max_poly_size = min(Max_Poly_Count,int(ptn_len/42))\n",
    "    \n",
    "    RNAcount = np.zeros(simtime+1)\n",
    "    \n",
    "    for i in range(configs):\n",
    "        \n",
    "        RnaName = 'R_' + locusNum\n",
    "        \n",
    "        t, Rcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RnaName))\n",
    "        \n",
    "        RNAcount = RNAcount + Rcount\n",
    "        \n",
    "        DegName = 'D_' + locusNum\n",
    "        \n",
    "        t, RDEGcount = trajs[i].getNumberTrajectory(species = trajs[i].species(DegName))\n",
    "    \n",
    "        RNAcount = RNAcount + RDEGcount\n",
    "        \n",
    "        if max_poly_size == 1:\n",
    "            \n",
    "            RiboID = 'RB_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RNAcount = RNAcount + RBcount\n",
    "            \n",
    "            RiboID = 'RBS_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RNAcount = RNAcount + RBcount\n",
    "            \n",
    "            RiboID = 'RBM_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RNAcount = RNAcount + RBcount\n",
    "            \n",
    "            RiboID = 'RBE_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RNAcount = RNAcount + RBcount\n",
    "            \n",
    "        else:\n",
    "            \n",
    "            RiboID = 'RB_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RNAcount = RNAcount + RBcount\n",
    "            \n",
    "            RiboID = 'PS_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RNAcount = RNAcount + RBcount\n",
    "            \n",
    "            RiboID = 'PS_' + locusNum + '_d'\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RNAcount = RNAcount + RBcount\n",
    "            \n",
    "            RiboID = 'PM_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RNAcount = RNAcount + RBcount\n",
    "            \n",
    "            RiboID = 'PM_' + locusNum + '_s'\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RNAcount = RNAcount + RBcount\n",
    "            \n",
    "            RiboID = 'PM_' + locusNum + '_d'\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RNAcount = RNAcount + RBcount\n",
    "            \n",
    "            RiboID = 'PE_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RNAcount = RNAcount + RBcount\n",
    "            \n",
    "            RiboID = 'PE_' + locusNum + '_s'\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RNAcount = RNAcount + RBcount\n",
    "        \n",
    "    \n",
    "    RNAcount = RNAcount/configs\n",
    "    \n",
    "    return t, RNAcount\n",
    "\n",
    "\n",
    "def pltAvgRibos(locusTag):\n",
    "    \n",
    "    locusNum = locusTag.split('_')[1].lstrip('0')\n",
    "    \n",
    "    rnasequence, aasequence = getSequences(locusTag)\n",
    "    \n",
    "    aaCount = defaultdict(int)\n",
    "    for aa in set(aasequence):\n",
    "        aaCount[aa] = aasequence.count(aa)\n",
    "    ptn_len = sum(list(aaCount.values()))\n",
    "\n",
    "    if ptn_len <= 125:\n",
    "        max_poly_size = 1\n",
    "    elif ptn_len > 125:\n",
    "        max_poly_size = min(Max_Poly_Count,int(ptn_len/42))\n",
    "    \n",
    "    RIBOcount = np.zeros(simtime+1)\n",
    "    \n",
    "    for i in range(configs):\n",
    "        \n",
    "        if max_poly_size == 1:\n",
    "            \n",
    "            RiboID = 'RB_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RIBOcount = RIBOcount + RBcount\n",
    "            \n",
    "            RiboID = 'RBS_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RIBOcount = RIBOcount + RBcount #+ RBcount\n",
    "            \n",
    "            RiboID = 'RBM_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RIBOcount = RIBOcount + RBcount \n",
    "            \n",
    "            RiboID = 'RBE_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RIBOcount = RIBOcount + RBcount\n",
    "            \n",
    "        else:\n",
    "            \n",
    "            RiboID = 'RB_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RIBOcount = RIBOcount + RBcount\n",
    "            \n",
    "            RiboID = 'PS_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RIBOcount = RIBOcount + RBcount #+ RBcount\n",
    "            \n",
    "            RiboID = 'PS_' + locusNum + '_d'\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RIBOcount = RIBOcount + RBcount\n",
    "            \n",
    "            RiboID = 'PM_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RIBOcount = RIBOcount + RBcount + RBcount\n",
    "            \n",
    "            RiboID = 'PM_' + locusNum + '_s'\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RIBOcount = RIBOcount + RBcount #+ RBcount\n",
    "            \n",
    "            RiboID = 'PM_' + locusNum + '_d'\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RIBOcount = RIBOcount + RBcount\n",
    "            \n",
    "            RiboID = 'PE_' + locusNum\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RIBOcount = RIBOcount + RBcount + RBcount\n",
    "            \n",
    "            RiboID = 'PE_' + locusNum + '_s'\n",
    "            \n",
    "            t, RBcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RiboID))\n",
    "            \n",
    "            RIBOcount = RIBOcount + RBcount \n",
    "\n",
    "    RIBOcount = RIBOcount / configs\n",
    "    \n",
    "    return RIBOcount\n",
    "\n",
    "def getRNAsequences(jcvi3AID):\n",
    "    # returns genomic and protein sequences\n",
    "    try:\n",
    "        rnasequence = genomeLocDict[jcvi3AID].extract(genome3A.seq).transcribe()\n",
    "        \n",
    "    except:\n",
    "        rnasequence = 0\n",
    "    \n",
    "    return rnasequence\n",
    "        \n",
    "# Create list of proteins with no proteomics data\n",
    "# ptnNoQuant = set()\n",
    "\n",
    "def getSequences(jcvi3AID):\n",
    "    # returns genomic and protein sequences\n",
    "    try:\n",
    "        rnasequence = genomeLocDict[jcvi3AID].extract(genome3A.seq).transcribe()\n",
    "        \n",
    "        # Using translation table 4 from NCBI: \"Mycoplasma Code\"\n",
    "        # https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi#SG4\n",
    "        aasequence  = genomeLocDict[jcvi3AID].extract(genome3A.seq).transcribe().translate(table=4)\n",
    "        \n",
    "    except:\n",
    "        aasequence  = 0\n",
    "        rnasequence = 0\n",
    "    \n",
    "    return rnasequence, aasequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calcAvgGeneReads(locusTag):\n",
    "    \n",
    "    locusNum = locusTag.split('_')[1].lstrip('0')\n",
    "    \n",
    "    rnasequence, aasequence = getSequences(locusTag)\n",
    "    \n",
    "    t, RNAcount = pltAvgRna(locusTag)\n",
    "    \n",
    "    numberMade = 0   \n",
    "    \n",
    "    for i in range(len(RNAcount)-1):\n",
    "        \n",
    "        if RNAcount[i] < RNAcount[i+1]:\n",
    "            \n",
    "            numberMade = numberMade + 1\n",
    "            \n",
    "    numberMade = numberMade / configs\n",
    "    \n",
    "    return numberMade\n",
    "\n",
    "\n",
    "def calcRepGeneReads(locusTag,rep):\n",
    "    \n",
    "    locusNum = locusTag.split('_')[1].lstrip('0')\n",
    "    \n",
    "    rnasequence, aasequence = getSequences(locusTag)\n",
    "    \n",
    "    aaCount = defaultdict(int)\n",
    "    for aa in set(aasequence):\n",
    "        aaCount[aa] = aasequence.count(aa)\n",
    "    ptn_len = sum(list(aaCount.values()))\n",
    "\n",
    "    if ptn_len <= 134:\n",
    "        max_poly_size = 1\n",
    "    elif ptn_len > 134:\n",
    "        max_poly_size = min(Max_Poly_Count,int(ptn_len/134))\n",
    "    \n",
    "    RNAcount = np.zeros(simtime+1)\n",
    "    i=rep-1\n",
    "        \n",
    "    RnaName = 'R_' + locusNum\n",
    "\n",
    "    t, Rcount = trajs[i].getNumberTrajectory(species = trajs[i].species(RnaName))\n",
    "\n",
    "    RNAcount = RNAcount + Rcount\n",
    "\n",
    "    DegName = 'D_' + locusNum\n",
    "\n",
    "    t, RDEGcount = trajs[i].getNumberTrajectory(species = trajs[i].species(DegName))\n",
    "\n",
    "    RNAcount = RNAcount + RDEGcount\n",
    "\n",
    "    for mp in range(Max_Poly_Count):\n",
    "\n",
    "        for pn in range(mp+1): #range(max_poly_size):\n",
    "\n",
    "            if (max_poly_size == 1) or (mp == 0):\n",
    "\n",
    "                Ribo_ID = 'RB' + str(mp+1) + '_' + locusNum + '_' + str(pn+1)\n",
    "\n",
    "                t, RRcount = trajs[i].getNumberTrajectory(species = trajs[i].species(Ribo_ID))\n",
    "\n",
    "                RNAcount = RNAcount + RRcount\n",
    "\n",
    "                break\n",
    "\n",
    "            elif max_poly_size > 1:\n",
    "\n",
    "                if pn == 0:\n",
    "\n",
    "                    Ribo_ID = 'RB' + str(mp+1) + '_' + locusNum + '_' + str(pn+1)\n",
    "\n",
    "                    t, RRcount = trajs[i].getNumberTrajectory(species = trajs[i].species(Ribo_ID))\n",
    "\n",
    "                    RNAcount = RNAcount + RRcount\n",
    "\n",
    "                elif (pn+1 == max_poly_size) or (pn == mp):\n",
    "\n",
    "                    Ribo_ID = 'RB' + str(mp+1) + '_' + locusNum + '_' + str(pn+1)\n",
    "\n",
    "                    t, RRcount = trajs[i].getNumberTrajectory(species = trajs[i].species(Ribo_ID))\n",
    "\n",
    "                    RNAcount = RNAcount + RRcount\n",
    "\n",
    "                    break\n",
    "\n",
    "                else:\n",
    "\n",
    "                    Ribo_ID = 'RB' + str(mp+1) + '_' + locusNum + '_' + str(pn+1)\n",
    "\n",
    "                    t, RRcount = trajs[i].getNumberTrajectory(species = trajs[i].species(Ribo_ID))\n",
    "\n",
    "                    RNAcount = RNAcount + RRcount\n",
    "    \n",
    "    numberMade = 0   \n",
    "    \n",
    "    for i in range(len(RNAcount)-1):\n",
    "        \n",
    "        if RNAcount[i] < RNAcount[i+1]:\n",
    "            \n",
    "            numberMade = numberMade + 1\n",
    "    \n",
    "    return numberMade\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load all necessary files\n",
    "# The reconstruction matches reactions with gene-protein-reactions (GPR) that use MMSYN1* IDs.\n",
    "reconstPD = pd.read_excel(\"../model_data/reconstruction.xlsx\", sheet_name='Reactions')\n",
    "\n",
    "# The annotation matches MMSYN1* IDs with JCVISYN3* IDs (or \"locus tags\").\n",
    "annotatPD = pd.read_excel(\"../model_data/FBA/Syn3A_annotation_compilation.xlsx\",\n",
    "                         sheet_name=\"Syn3A_annotation_compilation_condensed\")\n",
    "\n",
    "annotatPD_updated = pd.read_excel(\"../model_data/proteomics_annotations.xlsx\",sheet_name=\"Sheet1\")\n",
    "\n",
    "# The genome data matches \"locus tags\" with AOE* protein IDs.\n",
    "# It provides both the gene sequence, needed for transcription reactions in the ODE model,\n",
    "# and the protein sequence, needed for translation reactions in the model.\n",
    "# This is the NCBI Gene Bank-formated file (https://www.ncbi.nlm.nih.gov/nuccore/CP014992.1).\n",
    "\n",
    "genomeFile2 = '../model_data/syn2.gb'\n",
    "genome2 = next(SeqIO.parse(genomeFile2, \"gb\"))\n",
    "\n",
    "# This is the NCBI Gene Bank-formated file (https://www.ncbi.nlm.nih.gov/nuccore/CP016816.2).\n",
    "genomeFile3A = '../model_data/syn3A.gb'\n",
    "genome3A = next(SeqIO.parse(genomeFile3A, \"gb\"))\n",
    "\n",
    "# The proteomics matches AOE IDs with quantitative proteomics data.\n",
    "proteomPD = pd.read_excel(\"../model_data/proteomics.xlsx\", sheet_name=\"Proteomics\", skiprows=[0] )\n",
    "\n",
    "genome_syn3A = list(SeqIO.parse(genomeFile3A, \"genbank\"))\n",
    "dna3A = genome_syn3A[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "AOEtoJ2 = dict()\n",
    "J2toAOE = dict()\n",
    "genomeLocDict = dict()\n",
    "genomePtnLocDict = dict()\n",
    "genomeRnaLocDict = dict()\n",
    "Locus3A = []\n",
    "\n",
    "for f in genome2.features:\n",
    "    if f.type == \"CDS\":\n",
    "        JCVSYN2_tag = f.qualifiers['locus_tag'][0]\n",
    "        #print(JCVSYN2_tag)\n",
    "        # Not all entries have an AOE protein_id\n",
    "        if('protein_id' in f.qualifiers.keys()):\n",
    "            AOE_locus = f.qualifiers['protein_id'][0]\n",
    "            AOEtoJ2[AOE_locus] = JCVSYN2_tag\n",
    "            J2toAOE[JCVSYN2_tag] = AOE_locus\n",
    "#             genomeLocDict[JCVSYN2_tag] = f.location\n",
    "        else:\n",
    "            print(\"Locus \", JCVSYN2_tag, \" has no AOE id!\")\n",
    "    if f.type == \"rRNA\":\n",
    "        JCVSYN2_tag = f.qualifiers['locus_tag'][0]\n",
    "#         genomeLocDict[JCVSYN2_tag] = f.location\n",
    "    if f.type == \"tRNA\":\n",
    "        JCVSYN2_tag = f.qualifiers['locus_tag'][0]\n",
    "#         genomeLocDict[JCVSYN2_tag] = f.location\n",
    "        \n",
    "for f in genome3A.features:\n",
    "    if f.type == \"CDS\":\n",
    "        JCVSYN3A_tag = f.qualifiers['locus_tag'][0]\n",
    "        Locus3A.append(JCVSYN3A_tag)\n",
    "        #print(JCVSYN2_tag)\n",
    "        # Not all entries have an AOE protein_id\n",
    "        if('protein_id' in f.qualifiers.keys()):\n",
    "#             AOE_locus = f.qualifiers['protein_id'][0]\n",
    "#             AOEtoJ2[AOE_locus] = JCVSYN2_tag\n",
    "#             J2toAOE[JCVSYN2_tag] = AOE_locus\n",
    "            genomePtnLocDict[JCVSYN3A_tag] = f.location\n",
    "            genomeLocDict[JCVSYN3A_tag] = f.location\n",
    "        else:\n",
    "            print(\"Locus \", JCVSYN3A_tag, \" is pseudo.\")\n",
    "    if f.type == \"rRNA\":\n",
    "        JCVSYN3A_tag = f.qualifiers['locus_tag'][0]\n",
    "        Locus3A.append(JCVSYN3A_tag)\n",
    "        genomeRnaLocDict[JCVSYN3A_tag] = f.location\n",
    "        genomeLocDict[JCVSYN3A_tag] = f.location\n",
    "    if f.type == \"tRNA\":\n",
    "        JCVSYN3A_tag = f.qualifiers['locus_tag'][0]\n",
    "        Locus3A.append(JCVSYN3A_tag)\n",
    "        genomeRnaLocDict[JCVSYN3A_tag] = f.location\n",
    "        genomeLocDict[JCVSYN3A_tag] = f.location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def repFlux(rxnName,rep):\n",
    "    \n",
    "    numMin = int(simtime/60)\n",
    "    \n",
    "    flux_list = []\n",
    "    \n",
    "    for i in range(1,numMin):\n",
    "        \n",
    "        fluxFileMin = workDir + str(rep) + '/fluxes/fluxDF_' + str(i) + 'min.csv'\n",
    "\n",
    "        fluxDFMin = pd.read_csv(fluxFileMin,header=None)\n",
    "\n",
    "        for index,row in fluxDFMin.iterrows():\n",
    "            rxnID = row[1]\n",
    "            if rxnID == rxnName:\n",
    "                flux = row[2]\n",
    "                flux_list.append(flux)\n",
    "                \n",
    "    return flux_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# t, avgCnt = pltAvgRna('RNA_0910','Ribo_0910','Deg_0910')\n",
    "# plt.plot(t/60, avgCnt)\n",
    "# print(ht)\n",
    "specID = 'M_rRNA_16S_c'\n",
    "# specID = 'Degradosome'\n",
    "specID = 'M_pi_c'\n",
    "# specID = 'M_glntrna_c'\n",
    "# specID = 'M_glutrnagln_c'\n",
    "specID = 'ATP_mRNAdeg'\n",
    "specID = 'ATP_translat'\n",
    "# specID = 'D_60'\n",
    "# specID = 'M_glutrnagln_c'\n",
    "# specID = 'CellSA'\n",
    "plt.plot(ht/60,avgHybridCnt(specID))\n",
    "# print(np.average(avgHybridCnt(specID)[600:]))\n",
    "plt.xlabel('Time (min)')\n",
    "# plt.xlim(0,10)\n",
    "plt.ylabel('Counts')\n",
    "plt.ylim(0,10000)\n",
    "# plt.title(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "specID = 'M_atp_c'\n",
    "# specID = 'ATP_trsc'\n",
    "rep = 1\n",
    "plt.figure(figsize=(6,4))\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "# plt.plot(ht/60,repHybridCnt(specID,rep))\n",
    "plt.plot(ht/60,repHybridConc(specID,rep))\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Concentration (mM)')\n",
    "# plt.title(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "specID = 'M_datp_c'\n",
    "plt.figure(figsize=(6,4))\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "plt.plot(ht/60,avgHybridConc(specID))\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Concentration (mM)')\n",
    "plt.title(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "specIDs = ['M_atp_c','M_adp_c','M_gtp_c','M_ctp_c','M_utp_c']\n",
    "specIDs_paper = ['ATP','ADP','GTP','CTP','UTP']\n",
    "rep=0\n",
    "plt.rcParams.update({'font.size': 8})\n",
    "plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "fig = plt.figure(figsize=(3.3,2.5))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "for specID in specIDs:\n",
    "#     plt.plot(ht/60,repHybridConc(specID,rep))\n",
    "    ax.plot(ht/60,avgHybridConc(specID))\n",
    "ax.set_xlabel('Time (min)')\n",
    "ax.set_ylabel('Concentration (mM)')\n",
    "ax.set_ylim(-0.1,6.9)\n",
    "# plt.legend(specIDs,loc=2)\n",
    "ax.legend(specIDs_paper,ncol=2) #, bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "\n",
    "plt.show()\n",
    "# plt.title(specID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rxnID = 'PAPA'\n",
    "rep = 1\n",
    "plt.figure(figsize=(6,4))\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "plt.plot(ft,repFlux(rxnID,rep))\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Flux (mM/s)')\n",
    "plt.title(rxnID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "riboPtnMetDF = pd.read_csv(\"../model_data/ribo_protein_metabolites.csv\")\n",
    "# riboPtnMetDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8,6))\n",
    "inPlot = 0\n",
    "\n",
    "total_mRNA = 0\n",
    "\n",
    "for locusTag in genomePtnLocDict:\n",
    "#     locusNum = locusTag.split('_')[1].lstrip('0')\n",
    "#     rnaID = 'R_' + locusNum\n",
    "#     riboName = 'RB1_' + locusNum + '_1'\n",
    "#     degName = 'D_' + locusNum\n",
    "#     t, avgCnt = pltAvgRna(rnaID,riboName,degName)\n",
    "    t, avgCnt = pltAvgRna(locusTag)\n",
    "    total_mRNA = total_mRNA + avgCnt[-1]\n",
    "    plt.plot(t/60, avgCnt)\n",
    "    inPlot = inPlot + 1\n",
    "\n",
    "print(inPlot)\n",
    "print(total_mRNA)\n",
    "\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average mRNA Count')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "inPlot = 0\n",
    "\n",
    "total_mRNA = 0\n",
    "\n",
    "Deg_tot = np.zeros(1201)\n",
    "RNAP_tot = np.zeros(1201)\n",
    "Ribo_tot = np.zeros(1201)\n",
    "\n",
    "Deg_tot = np.zeros(simtime+1)\n",
    "RNAP_tot = np.zeros(simtime+1)\n",
    "Ribo_tot = np.zeros(simtime+1)\n",
    "\n",
    "for locusTag in genomePtnLocDict:\n",
    "#     rnaID = 'M_RNA_' + locusTag + '_c'\n",
    "    locusNum = locusTag.split('_')[1].lstrip('0')\n",
    "    rnapName = 'RP_' + locusNum\n",
    "#     riboName = 'RB_' + locusNum\n",
    "    degName = 'D_' + locusNum\n",
    "    t, rnapcnt = pltAvgSpec(rnapName)\n",
    "    t, degcnt = pltAvgSpec(degName)\n",
    "    ribocnt = pltAvgRibos(locusTag)\n",
    "#     t, ribocnt = pltAvgSpec(riboName)\n",
    "#     t, avgCnt = pltAvgRna(rnaID,riboName,degName)\n",
    "#     total_mRNA = total_mRNA + avgCnt[-1]\n",
    "#     plt.plot(t/60, avgCnt)\n",
    "#     inPlot = inPlot + 1\n",
    "    RNAP_tot = RNAP_tot + rnapcnt\n",
    "    Deg_tot = Deg_tot + degcnt\n",
    "    Ribo_tot = Ribo_tot + ribocnt\n",
    "\n",
    "# print(inPlot)\n",
    "# print(total_mRNA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(2.5,6/8*2.5))\n",
    "\n",
    "plt.rcParams.update({'font.size': 8})\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "plt.rcParams.update({'font.size': 8})\n",
    "\n",
    "ax.plot(t/60,Deg_tot,'k',linewidth=1)\n",
    "\n",
    "# print(np.average(Deg_tot[600:]))\n",
    "# print(np.std(Deg_tot[600:]))\n",
    "\n",
    "# ax.set_xlim(1,21)\n",
    "# ax.set_ylim(7,22)\n",
    "\n",
    "ax.set_xlabel('Time (min)')\n",
    "ax.set_ylabel('Active Degradosomes')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(2.5,6/8*2.5))\n",
    "\n",
    "ax = plt.gca()\n",
    "plt.rcParams.update({'font.size': 8})\n",
    "\n",
    "t,rnapol = pltAvgSpec('RNApol')\n",
    "\n",
    "ax.plot(t/60,RNAP_tot,'k',linewidth=1)\n",
    "\n",
    "ax.set_xlim(1,21)\n",
    "ax.set_ylim(50,70)\n",
    "    \n",
    "ax.set_xlabel('Time (min)')\n",
    "ax.set_ylabel('Active RNAP')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(2.5,6/8*2.5))\n",
    "\n",
    "ax = plt.gca()\n",
    "plt.rcParams.update({'font.size': 8})\n",
    "\n",
    "# t,rnapol = pltAvgSpec('RNApol')\n",
    "\n",
    "ax.plot(t/60,Ribo_tot,'k',linewidth=1)\n",
    "\n",
    "# ax.set_xlim(1,21)\n",
    "# ax.set_ylim(120,230)\n",
    "\n",
    "# print(np.average(Ribo_tot[600:]))\n",
    "\n",
    "ax.set_xlabel('Time (min)')\n",
    "ax.set_ylabel('Active Ribosomes')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metListmRNA = []\n",
    "\n",
    "# for rnaID in ModelSpecies:\n",
    "#     if \"RNA_MMS\" in rnaID:\n",
    "#         metListmRNA.append(rnaID)\n",
    "\n",
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "# time = np.arange(0,simtime+1,1)\n",
    "\n",
    "# for mmcode in gene_list_other:\n",
    "#     mRNA = \"M_RNA_\" + mmcode + \"_c\"\n",
    "#     metListmRNA.append(mRNA)\n",
    "#     plt.plot(time/60,AverageSpec(mRNA),'y')\n",
    "\n",
    "for index, row in riboPtnMetDF.iterrows():\n",
    "    mmcode = row[\"gene\"]\n",
    "    locusNum = mmcode.split('_')[1]\n",
    "    locusTag = 'JCVISYN3A_' + locusNum\n",
    "#     rnaID = 'M_RNA_' + locusTag + '_c'\n",
    "    rnaID = 'RNA_' + locusNum\n",
    "    riboName = 'Ribo_' + locusNum\n",
    "    degName = 'Deg_' + locusNum\n",
    "#     time, avgTrace = pltAvgRna(rnaID,riboName,degName)\n",
    "    time, avgTrace = pltAvgRna(locusTag)\n",
    "    \n",
    "    if avgTrace[-1]>2:\n",
    "        print(locusTag,avgTrace[-1])\n",
    "    metListmRNA.append(locusTag)\n",
    "     \n",
    "#     avgTrace = AverageSpec(mRNA)\n",
    "#     plt.plot(time/60,AverageSpec(mRNA),'crimson')\n",
    "    plt.plot(time/60,avgTrace,'crimson')\n",
    "    \n",
    "plt.ylim(0,3)\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average mRNA Count')\n",
    "    \n",
    "# metListPtns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "# time = np.arange(0,simtime+1,1)\n",
    "\n",
    "# metListmRNA.append(\"M_RNA_JCVISYN3A_0065_c\")\n",
    "# plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0065_c\"),'g')\n",
    "\n",
    "# metListmRNA.append(\"M_RNA_JCVISYN3A_0621_c\")\n",
    "# plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0621_c\"),'g')\n",
    "\n",
    "for locusTag in genomePtnLocDict:\n",
    "\n",
    "    locus_code = locusTag.split('3A_')[1]\n",
    "    mmcode = 'MMSYN1_' + locus_code\n",
    "    locusNum = mmcode.split('_')[1]\n",
    "    JCVI3AID = \"JCVISYN3A_\" + locusNum\n",
    "\n",
    "    function = annotatPD_updated.loc[ annotatPD_updated.iloc[:,3] == JCVI3AID ].iloc[0, 5].strip()\n",
    "    if 'Metabolism' in function:\n",
    "        \n",
    "#         print(JCVI3AID)\n",
    "\n",
    "        rnaID = 'RNA_' + locusNum\n",
    "        riboName = 'Ribo_' + locusNum\n",
    "        degName = 'Deg_' + locusNum\n",
    "#         time, avgTrace = pltAvgRna(rnaID,riboName,degName)\n",
    "        time, avgTrace = pltAvgRna(locusTag)\n",
    "        if avgTrace[-1]>2:\n",
    "            print(locusTag,avgTrace[-1])\n",
    "        metListmRNA.append(locusTag)\n",
    "\n",
    "        plt.plot(time/60,avgTrace,'g')\n",
    "\n",
    "# plt.xlim(0,20)\n",
    "plt.ylim(0,5)\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average mRNA Count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "# time = np.arange(0,simtime+1,1)\n",
    "\n",
    "# metListmRNA.append(\"M_RNA_JCVISYN3A_0065_c\")\n",
    "# plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0065_c\"),'g')\n",
    "\n",
    "# metListmRNA.append(\"M_RNA_JCVISYN3A_0621_c\")\n",
    "# plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0621_c\"),'g')\n",
    "\n",
    "for locusTag in genomePtnLocDict:\n",
    "\n",
    "    locus_code = locusTag.split('3A_')[1]\n",
    "    mmcode = 'MMSYN1_' + locus_code\n",
    "    locusNum = mmcode.split('_')[1]\n",
    "    JCVI3AID = \"JCVISYN3A_\" + locusNum\n",
    "\n",
    "    function = annotatPD_updated.loc[ annotatPD_updated.iloc[:,3] == JCVI3AID ].iloc[0, 5].strip()\n",
    "    if 'Metabolism' in function:\n",
    "        \n",
    "#         print(JCVI3AID)\n",
    "\n",
    "        rnaID = 'RNA_' + locusNum\n",
    "        riboName = 'Ribo_' + locusNum\n",
    "        degName = 'Deg_' + locusNum\n",
    "#         time, avgTrace = pltAvgRna(rnaID,riboName,degName)\n",
    "        time, avgTrace = pltAvgRna(locusTag)\n",
    "        if avgTrace[-1]>2:\n",
    "            print(locusTag,avgTrace[-1])\n",
    "        metListmRNA.append(locusTag)\n",
    "\n",
    "        plt.plot(time/60,avgTrace,'g')\n",
    "\n",
    "# plt.xlim(0,20)\n",
    "plt.ylim(0,5)\n",
    "ax = plt.gca()\n",
    "ax.axes.xaxis.set_visible(False)\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average mRNA Count')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metListmRNA = []\n",
    "plt.rcParams.update({'font.size': 8})\n",
    "\n",
    "fig = plt.figure(figsize=(3.3,2.5))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "# time = np.arange(0,simtime+1,1)\n",
    "\n",
    "# metListmRNA.append(\"M_RNA_JCVISYN3A_0065_c\")\n",
    "# plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0065_c\"),'g')\n",
    "\n",
    "# metListmRNA.append(\"M_RNA_JCVISYN3A_0621_c\")\n",
    "# plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0621_c\"),'g')\n",
    "\n",
    "for locusTag in genomePtnLocDict:\n",
    "\n",
    "    locus_code = locusTag.split('3A_')[1]\n",
    "    mmcode = 'MMSYN1_' + locus_code\n",
    "    locusNum = mmcode.split('_')[1]\n",
    "    JCVI3AID = \"JCVISYN3A_\" + locusNum\n",
    "\n",
    "    function = annotatPD_updated.loc[ annotatPD_updated.iloc[:,3] == JCVI3AID ].iloc[0, 5].strip()\n",
    "    if 'Genetic' in function:\n",
    "        \n",
    "#         print(JCVI3AID)\n",
    "\n",
    "        rnaID = 'RNA_' + locusNum\n",
    "        riboName = 'Ribo_' + locusNum\n",
    "        degName = 'Deg_' + locusNum\n",
    "#         time, avgTrace = pltAvgRna(rnaID,riboName,degName)\n",
    "        time, avgTrace = pltAvgRna(locusTag)\n",
    "        if avgTrace[-1]>2:\n",
    "            print(locusTag,avgTrace[-1])\n",
    "        metListmRNA.append(locusTag)\n",
    "\n",
    "        ax.plot(time/60,avgTrace,'dodgerblue')\n",
    "        \n",
    "# ax.set_xlim(0,20)\n",
    "ax.set_ylim(-0.1,5.1)\n",
    "ax.set_xlabel('Time (min)')\n",
    "ax.set_ylabel('Average mRNA Count')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "# time = np.arange(0,simtime+1,1)\n",
    "\n",
    "# metListmRNA.append(\"M_RNA_JCVISYN3A_0065_c\")\n",
    "# plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0065_c\"),'g')\n",
    "\n",
    "# metListmRNA.append(\"M_RNA_JCVISYN3A_0621_c\")\n",
    "# plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0621_c\"),'g')\n",
    "\n",
    "for locusTag in genomePtnLocDict:\n",
    "\n",
    "    locus_code = locusTag.split('3A_')[1]\n",
    "    mmcode = 'MMSYN1_' + locus_code\n",
    "    locusNum = mmcode.split('_')[1]\n",
    "    JCVI3AID = \"JCVISYN3A_\" + locusNum\n",
    "\n",
    "    function = annotatPD_updated.loc[ annotatPD_updated.iloc[:,3] == JCVI3AID ].iloc[0, 5].strip()\n",
    "    if 'Cellular Processes' in function:\n",
    "        \n",
    "#         print(JCVI3AID)\n",
    "\n",
    "        rnaID = 'RNA_' + locusNum\n",
    "        riboName = 'Ribo_' + locusNum\n",
    "        degName = 'Deg_' + locusNum\n",
    "#         time, avgTrace = pltAvgRna(rnaID,riboName,degName)\n",
    "        time, avgTrace = pltAvgRna(locusTag)\n",
    "        if avgTrace[-1]>2:\n",
    "            print(locusTag,avgTrace[-1])\n",
    "        metListmRNA.append(locusTag)\n",
    "\n",
    "        plt.plot(time/60,avgTrace,'darkorange')\n",
    "        \n",
    "plt.ylim(0,3)\n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average mRNA Count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.rcParams.update({'font.size': 18})\n",
    "\n",
    "plt.figure(figsize=(8,6))\n",
    "\n",
    "# time = np.arange(0,simtime+1,1)\n",
    "\n",
    "# metListmRNA.append(\"M_RNA_JCVISYN3A_0065_c\")\n",
    "# plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0065_c\"),'g')\n",
    "\n",
    "# metListmRNA.append(\"M_RNA_JCVISYN3A_0621_c\")\n",
    "# plt.plot(time/60,AverageSpec(\"M_RNA_JCVISYN3A_0621_c\"),'g')\n",
    "\n",
    "for locusTag in genomePtnLocDict:\n",
    "\n",
    "    locus_code = locusTag.split('3A_')[1]\n",
    "    mmcode = 'MMSYN1_' + locus_code\n",
    "    locusNum = mmcode.split('_')[1]\n",
    "    JCVI3AID = \"JCVISYN3A_\" + locusNum\n",
    "\n",
    "    function = annotatPD_updated.loc[ annotatPD_updated.iloc[:,3] == JCVI3AID ].iloc[0, 5].strip()\n",
    "    if 'Unclear' in function:\n",
    "        \n",
    "#         print(JCVI3AID)\n",
    "\n",
    "        rnaID = 'RNA_' + locusNum\n",
    "        riboName = 'Ribo_' + locusNum\n",
    "        degName = 'Deg_' + locusNum\n",
    "#         time, avgTrace = pltAvgRna(rnaID,riboName,degName)\n",
    "        time, avgTrace = pltAvgRna(locusTag)\n",
    "        if avgTrace[-1]>2:\n",
    "            print(locusTag,avgTrace[-1])\n",
    "        metListmRNA.append(locusTag)\n",
    "\n",
    "        plt.plot(time/60,avgTrace,'gray')\n",
    "\n",
    "plt.ylim(0,3)        \n",
    "plt.xlabel('Time (min)')\n",
    "plt.ylabel('Average mRNA Count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# inPlot = 0\n",
    "total_ptns = 0\n",
    "\n",
    "final_ptns = []\n",
    "final_ptns_scaled = []\n",
    "\n",
    "overTot = 0\n",
    "\n",
    "for locusTag in genomePtnLocDict:\n",
    "    locusNum = locusTag.split('_')[1].lstrip('0')\n",
    "    ptnID = 'P_' + locusNum\n",
    "#     locusNum = locusTag.split('_')[1]\n",
    "#     riboName = 'Ribo_' + locusNum\n",
    "    try:\n",
    "        t, avgCnt = pltAvgPtn(ptnID)\n",
    "    #     total_ptns = total_ptns + avgCnt[-1]\n",
    "        if avgCnt[0] >= 10:\n",
    "            scaledCnt = avgCnt[-1]/avgCnt[0]\n",
    "            cntMade = avgCnt[-1] - avgCnt[0]\n",
    "            total_ptns = total_ptns + cntMade\n",
    "            final_ptns.append(cntMade)\n",
    "            final_ptns_scaled.append(scaledCnt)\n",
    "            if scaledCnt > 1.4:\n",
    "                print(ptnID, ' : ', avgCnt[0], ' : ', scaledCnt)\n",
    "                overTot = overTot + 1\n",
    "            if scaledCnt < 1.05:\n",
    "                print(ptnID, ' : ', avgCnt[0], ' : ', scaledCnt)\n",
    "                overTot = overTot + 1\n",
    "                \n",
    "    except:\n",
    "        continue\n",
    "#     plt.plot(t/60, avgCnt)\n",
    "#     inPlot = inPlot + 1\n",
    "\n",
    "print(len(final_ptns))\n",
    "print(total_ptns)\n",
    "print(overTot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8,6))\n",
    "plt.hist(final_ptns, bins = 30)\n",
    "plt.xlabel('Number Generated in 20 min')\n",
    "plt.ylabel('Number of Unique Proteins')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 8})\n",
    "plt.rcParams.update({\"font.family\": 'sans-serif'})\n",
    "\n",
    "fig = plt.figure(figsize=(3.3,2.5))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "binwidth = 0.02\n",
    "\n",
    "ax.hist(final_ptns_scaled, bins=np.arange(min(final_ptns_scaled), max(final_ptns_scaled) + binwidth, binwidth),\n",
    "        edgecolor='k', color='lightgreen')# bins = 80)\n",
    "\n",
    "ax.axvline(x=np.average(final_ptns_scaled),c='b',lw=2,label='Average: '+str(np.round(np.average(final_ptns_scaled),2)))\n",
    "ax.axvline(x=np.median(final_ptns_scaled),c='r',lw=2,label='Median: '+str(np.round(np.median(final_ptns_scaled),2)))\n",
    "\n",
    "ax.set_xlabel('Scaled Protein Counts After 20 min')\n",
    "ax.set_ylabel('Number of Unique Proteins')\n",
    "ax.legend()\n",
    "ax.set_xlim(0.95,1.8)\n",
    "ax.set_ylim(0,90)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.average(final_ptns_scaled)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_ptns = 0\n",
    "\n",
    "# final_ptns = []\n",
    "# final_ptns_scaled = []\n",
    "\n",
    "avg_gene_reads = []\n",
    "avg_RNA_reads = []\n",
    "\n",
    "overTot = 0\n",
    "\n",
    "for locusTag in genomePtnLocDict:\n",
    "    locusNum = locusTag.split('_')[1].lstrip('0')\n",
    "    ptnID = 'P_' + locusNum\n",
    "#     locusNum = locusTag.split('_')[1]\n",
    "#     riboName = 'Ribo_' + locusNum\n",
    "    try:\n",
    "        t, avgCnt = pltAvgPtn(ptnID)\n",
    "    #     total_ptns = total_ptns + avgCnt[-1]\n",
    "        if avgCnt[0] >= 10:\n",
    "            scaledCnt = avgCnt[-1]/avgCnt[0]\n",
    "            cntMade = avgCnt[-1] - avgCnt[0]\n",
    "            total_ptns = total_ptns + cntMade\n",
    "#             final_ptns.append(cntMade)\n",
    "#             final_ptns_scaled.append(scaledCnt)\n",
    "            gene_reads = calcAvgGeneReads(locusTag)\n",
    "            time, avgRna = pltAvgRna(locusTag)\n",
    "            total_RNA = gene_reads + avgRna[0]\n",
    "            avg_gene_reads.append(gene_reads)\n",
    "            avg_RNA_reads.append(cntMade/total_RNA)\n",
    "            if scaledCnt > 1.4:\n",
    "                print(ptnID, ' : ', avgCnt[0], ' : ', scaledCnt)\n",
    "                overTot = overTot + 1\n",
    "            if scaledCnt == 1.0:\n",
    "                print(ptnID, ' : ', avgCnt[0], ' : ', scaledCnt)\n",
    "                overTot = overTot + 1\n",
    "                \n",
    "            if gene_reads > 10:\n",
    "                print(locusTag, ' gene reads : ', gene_reads)\n",
    "                \n",
    "            if gene_reads < 0.5:\n",
    "                print(locusTag, ' gene reads : ', gene_reads)\n",
    "                \n",
    "            if (cntMade/total_RNA > 6) or (cntMade/total_RNA < 3):\n",
    "                print(locusTag, ' mRNA reads : ', cntMade/total_RNA)\n",
    "                \n",
    "    except:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(min(avg_gene_reads))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(2.5,6/8*2.5))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "binwidth = 1\n",
    "ax.hist(avg_gene_reads, bins=np.arange(min(avg_gene_reads), max(avg_gene_reads) + binwidth, binwidth),\n",
    "        edgecolor='k', color='darkorange')\n",
    "ax.set_xlim(-0.5,20)\n",
    "ax.set_xlabel('Transcription Events/Protein-Coding Gene')\n",
    "ax.set_ylabel('Genes')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(2.5,6/8*2.5))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "binwidth=1\n",
    "ax.hist(avg_RNA_reads, bins=np.arange(min(avg_RNA_reads), max(avg_RNA_reads) + binwidth, binwidth),\n",
    "        edgecolor='k', color='r')# bins = 20)\n",
    "ax.set_xlabel('Proteins Translated per mRNA')\n",
    "ax.set_ylabel('Uniquie mRNA')\n",
    "# plt.xlim(0,8.5)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(min(avg_RNA_reads))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getRNAsequences(jcvi3AID):\n",
    "    # returns genomic and protein sequences\n",
    "    try:\n",
    "        rnasequence = genomeLocDict[jcvi3AID].extract(genome3A.seq).transcribe()\n",
    "        \n",
    "    except:\n",
    "        rnasequence = 0\n",
    "    \n",
    "    return rnasequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define how to calculate transcription rate constants as in equation 3 for transcription reactions.\n",
    "# Uses mature transcript length and proteomics for promoter strength.\n",
    "\n",
    "def TranscriptRate(rnaMetID, ptnMetID, rnasequence, jcvi2ID):\n",
    "    # Add trascription reaction\n",
    "    \n",
    "    # Check that we know all bases used in the sequence\n",
    "    if ( set(rnasequence) - set(baseMap.keys()) ):\n",
    "        raise Exception(\"Unknown base(s) in RNA sequence {}\".format(set(rnasequence) - set(baseMap.keys())) )\n",
    "    \n",
    "    # Count how many times each base is used\n",
    "    baseCount = defaultdict(int)\n",
    "    for base in set(rnasequence):\n",
    "        baseCount[base] = rnasequence.count(base)\n",
    "    \n",
    "\n",
    "    ptnCount, ptnName = getPtnCount(ptnMetID, jcvi2ID)\n",
    "    \n",
    "    kcat_mod = min(rnaPolKcat*(ptnCount/(180)),85)\n",
    "\n",
    "    kcat_mod = max(10,kcat_mod)\n",
    "    \n",
    "    # The rate form needs specific sequence data for the first two monomers:\n",
    "#     paramDict[\"CMonoA\"] = baseMap[ rnasequence[0] ]\n",
    "#     paramDict[\"CMonoB\"] = baseMap[ rnasequence[1] ]\n",
    "#     paramDict[\"KDA\"] = rnaPolKd # Since we are current;y using the same Kd for all nucleotides\n",
    "#     paramDict[\"KDB\"] = rnaPolKd \n",
    "    \n",
    "    # Add total number of monomers to parameter dict\n",
    "    \n",
    "    CMono1 = baseMap[ rnasequence[0] ]\n",
    "    \n",
    "    CMono2 = baseMap[ rnasequence[1] ]\n",
    "\n",
    "    n_tot = sum(list(baseCount.values()))\n",
    "\n",
    "    NMono_A = baseCount[\"A\"]\n",
    "    \n",
    "    NMono_U = baseCount[\"U\"]\n",
    "    \n",
    "    NMono_C = baseCount[\"C\"]\n",
    "    \n",
    "    NMono_G = baseCount[\"G\"]\n",
    "    \n",
    "    NMonoDict = [NMono_A,NMono_C,NMono_G,NMono_U]\n",
    "#     print(NMonoDict)\n",
    "    \n",
    "    \n",
    "    NMonoSum = NMono_A*rnaPolKd/ATPconc + NMono_C*rnaPolKd/CTPconc + NMono_U*rnaPolKd/UTPconc + NMono_G*rnaPolKd/GTPconc\n",
    "    \n",
    "\n",
    "    k_transcription = kcat_mod / ((rnaPolKd**2)/(CMono1*CMono2) + NMonoSum + n_tot - 1)\n",
    "    \n",
    "#     k_transcription = k_transcription * NaV\n",
    "    \n",
    "    return k_transcription\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ptnNoQuant = set()\n",
    "\n",
    "def getPtnCount(newMetID, jcvi2ID):\n",
    "    \n",
    "    # Check if protein quantification is available.\n",
    "    try:\n",
    "        if jcvi2ID.startswith(\"JCVIman_\"):\n",
    "            aoeID = manGPRPD.loc[ manGPRPD.MM == jcvi2ID.replace(\"JCVIman_\",\"\"), \"AOE\" ].values[0]\n",
    "        else:\n",
    "            aoeID = J2toAOE[ jcvi2ID ]\n",
    "        \n",
    "        ptnCount = max(defaultPtnCount,round(proteomPD.loc[ proteomPD.Protein == aoeID ].iloc[0,21]))\n",
    "#         \n",
    "        ptnName  = proteomPD.loc[ proteomPD.Protein == aoeID ].iloc[0,1].replace(\n",
    "            \" [synthetic bacterium JCVI-Syn3.0]\",\"\")\n",
    "        \n",
    "#         ptnConcentration = ptnCount*countToMiliMol\n",
    "    except:\n",
    "        print(\"WARNING: No protein count for\", newMetID)\n",
    "        print(\"Using default protein concentration.\")\n",
    "\n",
    "        ptnName = newMetID\n",
    "        ptnCount = defaultPtnCount\n",
    "#         ptnConcentration = defaultPtnConcentration\n",
    "\n",
    "        ptnNoQuant.add(newMetID)\n",
    "    \n",
    "    return ptnCount, ptnName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseMap = OrderedDict({ \"A\":\"M_atp_c\", \"U\":\"M_utp_c\", \"G\":\"M_gtp_c\", \"C\":\"M_ctp_c\" })\n",
    "# baseMapToMonoP = OrderedDict({ \"A\":\"M_amp_c\", \"U\":\"M_ump_c\", \"G\":\"M_gmp_c\", \"C\":\"M_cmp_c\" })\n",
    "\n",
    "# Global parameters for transcription\n",
    "rnaPolKcat = 20 # nt/s\n",
    "rnaPolK0 = 1e-4 #mM\n",
    "rnaPolKd = 0.1 #mM\n",
    "\n",
    "rrnaPolKcat = 85 # nt/s\n",
    "\n",
    "krnadeg = 0.00578/2 # 1/s\n",
    "# rna_deg_rate = sim.rateConst('RNAdeg', krnadeg, 2)\n",
    "\n",
    "ptnDegRate = 7.70e-06 # 1/s\n",
    "\n",
    "ATPconc = 4 #mM\n",
    "UTPconc = 1 #mM\n",
    "CTPconc = 1 #mM\n",
    "GTPconc = 2 #mM\n",
    "\n",
    "# Cell radius (meters):\n",
    "# r_cell = 2.5*(10**-7)\n",
    "r_cell = 2.0*(10**-7) # m\n",
    "\n",
    "CytoVolume = (4*np.pi/3)*1000*r_cell**3 # L\n",
    "cellVolume = CytoVolume\n",
    "\n",
    "subvolume_vol = 1000*(8e-9)**3 # L\n",
    "\n",
    "# print(cellVolume)\n",
    "\n",
    "# Avogadro:\n",
    "avgdr   = 6.022e23 # molec/mol\n",
    "Avognum = avgdr\n",
    "\n",
    "NaV = Avognum * subvolume_vol\n",
    "\n",
    "countToMiliMol = 1000/(avgdr*cellVolume)\n",
    "\n",
    "defaultPtnCount = 10\n",
    "\n",
    "RnaPconc = 187*countToMiliMol # mM\n",
    "\n",
    "# Global parameter for degradation of mRNAs\n",
    "rnaDegRate = 0.00578/2 # 1/s\n",
    "\n",
    "degrad_bind_rate = 11/60/countToMiliMol*1000 #1/M/s\n",
    "# degrad_bind_rate = 11/60*NaV #1/M/s\n",
    "# deg_bind_rate = sim.rateConst('RNAdeg', degrad_bind_rate, 2)\n",
    "\n",
    "# Create a map for rna sequence to NTP concentration.\n",
    "baseMap = OrderedDict({ \"A\":ATPconc, \"U\":UTPconc, \"G\":GTPconc, \"C\":CTPconc })\n",
    "\n",
    "# Create Dictionaries to map tRNAs to associated aa abbreviations in protein sequences.\n",
    "aaMap = OrderedDict({\"A\":\"M_ala__L_c\", \"R\":\"M_arg__L_c\", \n",
    "    \"N\":\"M_asn__L_c\", \"D\":\"M_asp__L_c\", \"C\":\"M_cys__L_c\", \"E\":\"M_glu__L_c\", \"Q\":\"M_gln__L_c\", \"G\":\"M_gly_c\", \n",
    "    \"H\":\"M_his__L_c\", \"I\":\"M_ile__L_c\", \"L\":\"M_leu__L_c\", \"K\":\"M_lys__L_c\", \"M\":\"M_met__L_c\", \"F\":\"M_phe__L_c\", \n",
    "    \"P\":\"M_pro__L_c\", \"S\":\"M_ser__L_c\", \"T\":\"M_thr__L_c\", \"W\":\"M_trp__L_c\", \"Y\":\"M_tyr__L_c\", \"V\":\"M_val__L_c\",\n",
    "    \"*\":\"Stop_Codon\"})\n",
    "\n",
    "aaTRNAMap = OrderedDict({\"A\":\"M_alatrna_c\", \"R\":\"M_argtrna_c\", \n",
    "    \"N\":\"M_asntrna_c\", \"D\":\"M_asptrna_c\", \"C\":\"M_cystrna_c\", \"E\":\"M_glutrna_c\", \"Q\":\"M_glntrna_c\", \"G\":\"M_glytrna_c\", \n",
    "    \"H\":\"M_histrna_c\", \"I\":\"M_iletrna_c\", \"L\":\"M_leutrna_c\", \"K\":\"M_lystrna_c\", \"M\":\"M_mettrna_c\", \"F\":\"M_phetrna_c\", \n",
    "    \"P\":\"M_protrna_c\", \"S\":\"M_sertrna_c\", \"T\":\"M_thrtrna_c\", \"W\":\"M_trptrna_c\", \"Y\":\"M_tyrtrna_c\", \"V\":\"M_valtrna_c\"})\n",
    "\n",
    "aaTRNAFreeMap = OrderedDict({\"A\":\"M_trnaala_c\", \"R\":\"M_trnaarg_c\", \n",
    "    \"N\":\"M_trnaasn_c\", \"D\":\"M_trnaasp_c\", \"C\":\"M_trnacys_c\", \"E\":\"M_trnaglu_c\", \"Q\":\"M_trnagln_c\", \"G\":\"M_trnagly_c\", \n",
    "    \"H\":\"M_trnahis_c\", \"I\":\"M_trnaile_c\", \"L\":\"M_trnaleu_c\", \"K\":\"M_trnalys_c\", \"M\":\"M_trnamet_c\", \"F\":\"M_trnaphe_c\", \n",
    "    \"P\":\"M_trnapro_c\", \"S\":\"M_trnaser_c\", \"T\":\"M_trnathr_c\", \"W\":\"M_trnatrp_c\", \"Y\":\"M_trnatyr_c\", \"V\":\"M_trnaval_c\"})\n",
    "\n",
    "\n",
    "# Global parameters for translation\n",
    "riboKcat = 12 # 1/s\n",
    "riboK0 = 4*25e-6 # mM\n",
    "riboKd = 0.0001 # mM\n",
    "\n",
    "ribosomeConc = 500*countToMiliMol # mM\n",
    "\n",
    "# Concentration of charged tRNA\n",
    "ctRNAconc = 150*countToMiliMol # mM\n",
    "\n",
    "# Global parameter for degradation of proteins\n",
    "# Derived from eLife's model, using average protein half life of 25 hours. \n",
    "ptnDegRate = 7.70e-06 # 1/s\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define how to add the particles and reactions for each protein and its corresponding mRNA and gene.\n",
    "\n",
    "def calc_mRNA_thalf(jcvi3AID):\n",
    "    locusNum = jcvi3AID.split('_')[1]\n",
    "    mmcode = 'MMSYN1_' + locusNum\n",
    "    # Checks if a translation to JCVISYN2* code is available\n",
    "    try:\n",
    "        jcvi2ID = annotatPD.loc[ annotatPD.iloc[:,5] == mmcode ].iloc[0, 13].strip()\n",
    "    except:\n",
    "        jcvi2ID = \"JCVIunk_\" + mmcode\n",
    "\n",
    "#     print(mmcode, jcvi2ID, jcvi3AID)\n",
    "    \n",
    "#     genes_in_model.append(jcvi3AID)\n",
    "    \n",
    "    # We name proteins after their locus tag from the gen bank entry to be M_PTN_JCVISYN3A_XXXX_c.\n",
    "    ptnMetID = 'PTN_' + jcvi3AID\n",
    "    \n",
    "    # If the protein is not in the model, add it:\n",
    "\n",
    "#     ModelSpecies.append(ptnMetID)\n",
    "\n",
    "    ptnCount, ptnName = getPtnCount(ptnMetID, jcvi2ID)\n",
    "#     print(ptnCount)\n",
    "    \n",
    "#     ptnCounts.append(ptnCount)\n",
    "\n",
    "#     print(ptnMetID, ptnCount)\n",
    "\n",
    "    geneMetID = jcvi3AID + '_gene'\n",
    "\n",
    "    # Get nucleotide and amino acid sequences, if available\n",
    "    rnasequence = getRNAsequences(jcvi3AID) #, aasequence\n",
    "#     print(rnasequence)\n",
    "#     print(aasequence)\n",
    "\n",
    "    if (rnasequence != 0): #and (aasequence != 0):\n",
    "        \n",
    "        locusNum = jcvi3AID.split('_')[1].lstrip('0')\n",
    "\n",
    "        rnaMetID = \"RNA_\" + locusNum\n",
    "        rnaName = \"(mRNA) \" + ptnName\n",
    "\n",
    "#         ModelSpecies.append(rnaMetID)\n",
    "    \n",
    "        species = []\n",
    "        species = [geneMetID, rnaMetID, ptnMetID]\n",
    "        \n",
    "        RnapName = 'RP_' + locusNum\n",
    "\n",
    "        riboName = 'Ribo_' + locusNum\n",
    "        degName = 'Deg_' + locusNum\n",
    "        \n",
    "        trsc_rate = TranscriptRate(rnaMetID, ptnMetID, rnasequence, jcvi2ID)\n",
    "        \n",
    "#         time, avgTrace = pltAvgRna(rnaMetID,riboName,degName)\n",
    "        time, avgTrace = pltAvgRna(jcvi3AID)\n",
    "        \n",
    "        t2, avgRnap = pltAvgSpec(RnapName)\n",
    "        \n",
    "        total_et = 0\n",
    "        total_messenger = 0\n",
    "\n",
    "        for i in range(int(len(avgTrace)/2),len(avgTrace)):\n",
    "#             print(i)\n",
    "            total_et = total_et + avgRnap[i]\n",
    "            total_messenger = total_messenger + avgTrace[i]\n",
    "            \n",
    "        avgET = total_et/(len(avgTrace) - len(avgTrace)/2)\n",
    "        avgMess = total_messenger/(len(avgTrace) - len(avgTrace)/2)\n",
    "        \n",
    "        kDeg = trsc_rate * avgET / avgMess\n",
    "        \n",
    "        tHalf = np.log(2)/kDeg/60\n",
    "        \n",
    "        return tHalf\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calc_mRNA_thalf('JCVISYN3A_0779')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "half_lives = []\n",
    "\n",
    "for locusTag in genomePtnLocDict:\n",
    "    thalf = calc_mRNA_thalf(locusTag)\n",
    "    half_lives.append(thalf)\n",
    "    if thalf>4 or thalf<0.1:\n",
    "        print(locusTag,thalf)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "half_lives_2 = []\n",
    "\n",
    "for half_life in half_lives:\n",
    "    if half_life and (half_life != 0) and (half_life<20):\n",
    "        half_lives_2.append(half_life)\n",
    "        if half_life < 0.16:\n",
    "            print(half_life)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 8})\n",
    "\n",
    "# fig = plt.figure(figsize=(2.5,6/8*2.5))\n",
    "fig = plt.figure(figsize=(3.3,2.5))\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "binwidth=0.5\n",
    "ax.hist(half_lives_2,\n",
    "         bins=np.arange(min(half_lives_2), max(half_lives_2) + binwidth, binwidth),\n",
    "         edgecolor='k',color='y')#bins=25)\n",
    "ax.axvline(x=np.average(half_lives_2),c='b',lw=2,label='Average: '+str(np.round(np.average(half_lives_2),2))+' min')\n",
    "ax.axvline(x=np.median(half_lives_2),c='r',lw=2,label='Median: '+str(np.round(np.median(half_lives_2),2))+' min')\n",
    "ax.set_xlabel('mRNA Half Life (min)')\n",
    "ax.set_ylabel('Number of Unique mRNA')\n",
    "ax.set_xlim(-0.25,17)\n",
    "ax.set_ylim(0,100)\n",
    "ax.legend()\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# plt.xscale('log')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.average(half_lives_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
